### DISCUSSION TITLE
I never worked on a VS Code extension, How to do that here?

### DISCUSSION THREAD

**Discussion by ewerybody** [mark-wiemer/ahkpp#42] [Q&A] [answered]
Hey! Is it OK if we gather some info on how to contribute to this project? I'd really like to learn :)

So. I got myself a clone with Github Desktop and said "I'd like to contribute".
Is that already good or do I NEED to have a proper fork?

Then: Let's take a simpe seeming example: I'd like to improve the **Formatter** How do I start this, can I use VS Code while working on VS Code ?!? 🤷‍♀️

**Comment by mark-wiemer** [comment, marked as answer]
I have been meaning to improve contributing guidelines, thank you for nudging me. :)

Of course you can use VS Code while working on VS Code! It's a marvelous tool. The first tutorial I would recommend is the [official extension authoring guide](https://code.visualstudio.com/api). This project uses the Extension API to hook into a lot of VS Code variables and keyboard shortcuts, so it's valuable to understand the basic structure of a VS Code extension before contributing to one. There you will learn how to work on an extension and see your changes (hint: it's as simple as pressing F5!).

After that, I'd create a fork of this repo and start work on a feature branch. If you don't know the basics of git/GitHub (working tree vs index, branches, pushing/fetching/pulling, resetting and amending commits, keeping commits atomic, etc.), I recommend walking through the [git docs](https://git-scm.com/docs), at least through "Inspection and Comparison". I know it might look like a lot, but take it one page at a time, and really try to understand what's going on. Try to predict what will happen before you execute a command. I made a lot of mistakes for the first few years while learning git, but it's super worth it in the long run. It's one of few tools that's omnipresent in software.

The formatting logic is done in [src/providers/formattingProvider.ts](https://github.com/mark-wiemer/vscode-autohotkey-plus-plus/blob/0a8138305a9db7ad4b14008a5ce4595bdeda49cd/src/providers/formattingProvider.ts#L33). It's a cryptic function that I'm still trying to fully decipher, so good luck! I'd recommend exploring the rest of the repo as well to get a feel for where everything is.

Once you have something you like, feel free to open a PR (remember, you'll need a fork for this). If you have ideas for how to fix issues, but no code to commit, you can leave comments on those issues to help others who may be working on them.

Hopefully that's a good start! I will be adding documentation to the codebase over the course of this month so we have everything we need in one place.
